rust/ql/lib/codeql/rust/controlflow/internal/Completion.qll

```diff
@@ -117,51 +117,63 @@ class BooleanCompletion extends ConditionalCompletion, TBooleanCompletion {
   override string toString() { result = "boolean(" + value + ")" }
 }
 
-/** Holds if `pat` is guaranteed to match at the point in the AST where it occurs. */
-pragma[nomagic]
-private predicate isExhaustiveMatch(Pat pat) {
-  (
-    pat instanceof WildcardPat
-    or
-    pat = any(IdentPat ip | not ip.hasPat() and ip = any(Variable v).getPat())
-    or
-    pat instanceof RestPat
-    or
-    // `let` statements without an `else` branch must be exhaustive
-    pat = any(LetStmt let | not let.hasLetElse()).getPat()
-    or
-    // `match` expressions must be exhaustive, so last arm cannot fail
-    pat = any(MatchExpr me).getLastArm().getPat()
-    or
-    // macro invocations are exhaustive if their expansion is
-    pat = any(MacroPat mp | isExhaustiveMatch(mp.getMacroCall().getExpanded()))
-    or
-    // parameter patterns must be exhaustive
-    pat = any(Param p).getPat()
-  ) and
-  not pat = any(ForExpr for).getPat() // workaround until `for` loops are desugared
+/**
+ * Holds if `pat` can not _itself_ be the cause of a pattern match failure. This
+ * does not mean that `pat` is irrefutable, as its children might be the cause
+ * of a failure.
+ */
+private predicate canCauseMatchFailure(Pat pat) {
+  pat instanceof LiteralPat
   or
-  exists(Pat parent | isExhaustiveMatch(parent) |
-    pat = parent.(BoxPat).getPat()
-    or
-    pat = parent.(IdentPat).getPat()
-    or
-    pat = parent.(MacroPat).getMacroCall().getExpanded()
-    or
-    pat = parent.(ParenPat).getPat()
-    or
-    pat = parent.(RecordPat).getRecordPatFieldList().getField(_).getPat()
-    or
-    pat = parent.(RefPat).getPat()
-    or
-    pat = parent.(TuplePat).getAField()
+  // NOTE: a `TupleStructPat` can cause a failure if it resolves to a an enum
+  // variant but not when it resolves to a tuple struct.
+  pat instanceof TupleStructPat
```

Comment:
`RecordPat` also needs to be added; enum variants are allowed to use record fields.